src/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.cpp

```diff
@@ -315,27 +361,57 @@ Propagator Propagator::FromEnvironment(const NumericalSolver& aNumericalSolver,
     };
 }
 
-void Propagator::registerDynamicsContext(const Shared<Dynamics>& aDynamicsSPtr)
+void Propagator::validateDynamicsSet() const
 {
-    // Store read coordinate subsets information
-    Array<Pair<Index, Size>> readInfo = Array<Pair<Index, Size>>::Empty();
-    readInfo.reserve(aDynamicsSPtr->getReadCoordinateSubsets().getSize());
-    for (const Shared<const CoordinateSubset>& subset : aDynamicsSPtr->getReadCoordinateSubsets())
+    Map<String, Size> currentDynamicsSets = {
```

Comment:
...continuing from the first comment of the review:

1. Just a quick thing, if you need to use the type of instances for your logic, you might want to use `type_index`, which should be cleaner (imo) and you don't need to "hardcode" class names. For example:

```cpp
#include <typeindex>

std::map<std::type_index, int> typeCount;
```

Then you can keep track of number of instance types:

```cpp
for (const auto& dynamics : this->dynamics) {
  ++typeCount[typeid(*dynamics)];
}
```

I had to search the syntax online, but the idea is not to hardcode class names (as those can change) but use Types instead.

2. This is the actual continuation of the first comment in the review: I think rather than looping and trying to find a common rule-set, maybe we can just implement here the logic. It's ok if we can't loop over a map, this is not to be re-used and you can have something like this (be warned this is pseudo-code):

```
// Keep track of celestial bodies for which gravity has been considered
gravityCelestialBodies = []


// Check there is only one CentralBodyGravity
if typeCouns[typeid(CentralBodyGravity)] != 1 {
  throw error...
} 

// Find and add the celestial body of the CentralBodyGravity
gravityCelestialBodies.add(centralBodyGravity.getCelestialBody())

// Check that the atmosphere, if any is the same as the Gravity Central Body
if centralBodyGravity.getCelestialBody() != atmosphericDrag.getCelestialBody() {
  throw error...
}

// Check there is one ThrusterDynamics mas
if typeCounts[typeid(ThrusterDynamics)] > 1 {
   throw error...
}

// Check no duplications amongst gravity bodys
for thirdBodyGravity in Dynacmis {
  if thirdBodyGravity.getCelestialBody() in gravityCelestialBodies {
    throw error...
  } else {
    gravityCelestialBodies.add(thirdBodyGravity.getCelestialBody();
  }
}

// Etc.
```

Anyway, not saying we need to do all of this now, that was just an example of ruleset. The message was that the ruleset is going to be heterogeneous enough for us to simply code it. 

⚠️ Also -> keep in mind that `Dynamics` abstract class was designed so that users can write their own implementation and simply provide it to the propagator and make full use of the rest of astro. This is just a reminder that we can validate as much as we want, but it's not 100% guaranteed that the dynamics set makes sense. `Propagator` is probably to be moved to Physics in the future, together with `Dynamics`. The validation we are doing here is astro-related, which is fine for now, and will probably be kept within the astro package in the future.

Reply:
Just an idea, if it becomes unmanageable, then we can always revert 👍

Reply:
I did what you suggested and its not super clean, and it also doesn't fully cover all the cases above, but that's fine because in some cases you can't even create a ThirdBodyDynamics for Earth for example.

Reply:
But I think its definitely better than what I had before and its good enough for now. Lmk what you think